flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/JdbcXaSinkFunction.java

```diff
@@ -0,0 +1,367 @@
+    // hanging XIDs - used for cleanup
+    // it's a list to support retries and scaling down
+    // possible transaction states: active, idle, prepared
+    // last element is the current xid
+    private transient Deque<Xid> hangingXids = new LinkedList<>();
+    private transient Xid currentXid;
+
+    /**
+     * Creates a {@link JdbcXaSinkFunction}.
+     *
+     * <p>All parameters must be {@link java.io.Serializable serializable}.
+     *
+     * @param xaFacade {@link XaFacade} to manage XA transactions
+     */
+    public JdbcXaSinkFunction(
+            String sql,
+            JdbcStatementBuilder<T> statementBuilder,
+            XaFacade xaFacade,
+            JdbcExecutionOptions executionOptions,
+            JdbcExactlyOnceOptions options) {
+        this(
+                new JdbcBatchingOutputFormat<>(
+                        xaFacade,
+                        executionOptions,
+                        context -> {
+                            Preconditions.checkState(
+                                    !context.getExecutionConfig().isObjectReuseEnabled(),
+                                    "objects can not be reused with JDBC sink function");
+                            return JdbcBatchStatementExecutor.simple(
+                                    sql, statementBuilder, Function.identity());
+                        },
+                        JdbcBatchingOutputFormat.RecordExtractor.identity()),
+                xaFacade,
+                XidGenerator.semanticXidGenerator(),
+                new XaSinkStateHandlerImpl(),
+                options,
+                new XaGroupOpsImpl(xaFacade));
+    }
+
+    /**
+     * Creates a {@link JdbcXaSinkFunction}.
+     *
+     * <p>All parameters must be {@link java.io.Serializable serializable}.
+     *
+     * @param format {@link JdbcBatchingOutputFormat} to write records with
+     * @param xaFacade {@link XaFacade} to manage XA transactions
+     * @param xidGenerator {@link XidGenerator} to generate new transaction ids
+     */
+    public JdbcXaSinkFunction(
+            JdbcBatchingOutputFormat<T, T, JdbcBatchStatementExecutor<T>> format,
+            XaFacade xaFacade,
+            XidGenerator xidGenerator,
+            XaSinkStateHandler stateHandler,
+            JdbcExactlyOnceOptions options,
+            XaGroupOps xaGroupOps) {
+        this.xaFacade = Preconditions.checkNotNull(xaFacade);
+        this.xidGenerator = Preconditions.checkNotNull(xidGenerator);
+        this.format = Preconditions.checkNotNull(format);
+        this.stateHandler = Preconditions.checkNotNull(stateHandler);
+        this.options = Preconditions.checkNotNull(options);
+        this.xaGroupOps = xaGroupOps;
+    }
+
+    @Override
+    public void initializeState(FunctionInitializationContext context) throws Exception {
+        JdbcXaSinkFunctionState state = stateHandler.load(context);
+        hangingXids = new LinkedList<>(state.getHanging());
```

Comment:
`Deque` is used to keep the current `Xid` the last; otherwise the order is not important.

>  the state recovery doesn't retain the order.

Could you explain what do you mean?